src/rapids_pre_commit_hooks/copyright.py

```diff
@@ -184,31 +184,39 @@ def get_target_branch_upstream_commit(repo, args):
         except ValueError:
             return None
 
+    commits_to_try = []
```

Comment:
Could you add some comments over each section of this `get_target_branch_upstream_commit()`?

I'm struggling follow with the nested try-catches and multiple points of early return.

Here's my interpretation, not sure if it's correct:

```python
def get_target_branch_upstream_commit(repo, args):
    # if the target branch can't be determined, target_commit is HEAD of the current branch on the local repo
    ...
    
    commits_to_try = []
    try:
        # check if the target branch exists locally
        target_branch = repo.heads[target_branch_name]
    except IndexError:
        # if the target branch doesn't exist locally, fetch the commit ID of its latest commit
        # on every remote in 'git remote' that has it
        ...
    else:
        # if the target branch DOES exist locally, consider the commit ID of its latest commit
        commits_to_try.append(target_branch.commit)
        # if the local branch is tracking a branch on a remote, consider the commit ID of the
        # latest commit on that remote branch
        target_branch_upstream = target_branch.tracking_branch()
        if target_branch_upstream:
            commits_to_try.append(target_branch_upstream.commit)

    # if any of those approaches identified commits to consider, return the commit ID
    # of the most recent one
    if commits_to_try:
        return max(commits_to_try, key=lambda commit: commit.committed_datetime)

    # if all that searching by target branch name failed, then
    # target_commit is HEAD of the current branch on the local repo
    try:
        return repo.head.commit
    except ValueError:
        return None
```

I don't understand what mix of circumstances could lead to that final `return None` being reached.

Reply:
Also... does the `repo.heads` object support `in`? If it does, I think this would be easier to understand if refactored like this:

```python
if target_branch_name in repo.heads:
    ....
else:
    ...
```

Instead of splitting the "the repo has a local `target_branch`" handling across the combination of the `try:` and later `else:` block.

Reply:
> does the `repo.heads` object support `in`?

One of the principles of Python is EAFP (easier to ask for forgiveness than permission), as opposed to LBYL (look before you leap) which you see in languages like C++. Python prefers to attempt to access an index and then do something else if an error is thrown.

Reply:
Your analysis is correct, but I realized there was a slight regression when the local branch doesn't exist - it won't search for a remote branch of the same name. I've just fixed this.

Reply:
I find this:

```python
try:
    # beginning of logic for situation A
except IndexError:
    # logic for situation B
else:
    # rest of logic for situation A
```

...more difficult to follow than this:

```python
if something:
    # logic for situation A
else:
    # logic for situation B
```

But don't let me hold up this PR over a matter of personal preference. Thanks anyway for considering it.

I would really appreciate comments similar to the ones I proposed above, though. If you don't want to add those here yourself I'd be happy to put up a separate PR proposing them.

Reply:
I added some explanation comments

Reply:
thank you!

Reply:
Tiny comment on this topic (just informational, not requesting changes). While you’re right that Python _tends_ to prefer EAFP, “EAFP over LBYL” is not a hard and fast rule in the language design. There are certainly cases where conditional logic is cleaner — but also LBYL can be faster in some cases! It depends on a lot of factors (data distribution, how often each path is taken, the cost of the conditional, etc), so pragmatism is usually better than strictly adhering to that “EAFP over LBYL” principle.